students/DiannaTingg/lessons/lesson03/assignment/test_basic_operations.py

```diff
@@ -0,0 +1,190 @@
+    bo.add_customer(**customer1)
+
+    with pytest.raises(pw.IntegrityError):
+        bo.add_customer(**customer1)
+
+    clear_database()
+
+
+def create_sample_database():
+    clear_database()
+
+    cm.database.create_tables([
+        cm.Customer
+    ])
+    bo.add_customer(**customer1)
+    bo.add_customer(**customer2)
+    bo.add_customer(**customer3)
+    bo.add_customer(**customer4)
+
+    cm.database.close()
+
+
+def test_search_customer():
+    create_sample_database()
+
+    test_customer = bo.search_customer("3")
+
+    assert test_customer["email_address"] == customer3["email_address"]
+
+    clear_database()
+
+
+def test_search_customer_missing():
+    create_sample_database()
+
+    assert bo.search_customer("8") == dict()
+
+    clear_database()
+
+
+def test_delete_customer():
+    create_sample_database()
+
+    number_of_customer2s = (cm.Customer.select().where(cm.Customer.customer_id == "2").count())
+    assert number_of_customer2s == 1
+
+    bo.delete_customer("2")
+    new_number_of_customer2s = (cm.Customer.select().where(cm.Customer.customer_id == "2").count())
+    assert new_number_of_customer2s == 0
+
+    clear_database()
+
+
+def test_delete_customer_missing():
+    create_sample_database()
+
+    number_of_customers = (cm.Customer.select().count())
+    assert number_of_customers == 4
+
+    bo.delete_customer("9")
+
+    new_number_of_customers = (cm.Customer.select().count())
+    assert new_number_of_customers == 4
+
+    clear_database()
+
+
+def test_update_customer_credit():
+    create_sample_database()
+
+    bo.update_customer_credit("1", 5000)
+
+    target_customer = cm.Customer.get(cm.Customer.customer_id == 1)
+
+    assert target_customer.credit_limit == 5000
+
+    clear_database()
+
+
+def test_update_customer_credit_missing():
+    create_sample_database()
+
+    with pytest.raises(pw.DoesNotExist):
+        bo.update_customer_credit("44", 300)
+
+    clear_database()
+
+
+def test_list_active_customers():
+    create_sample_database()
+
+    assert bo.list_active_customers() == 2
+
+    clear_database()
+
+
+def test_list_active_customers_empty():
+    create_empty_database()
+
+    assert bo.list_active_customers() == 0
+
+    clear_database()
+
+
+def test_integration():
+    create_empty_database()
+
+    # Add some customers
+    bo.add_customer(**customer1)
+    bo.add_customer(**customer2)
+    bo.add_customer(**customer3)
+    bo.add_customer(**customer4)
+
+    # Delete one customer
+    bo.delete_customer(customer3["customer_id"])
+
+    # Update customer credit
+    bo.update_customer_credit("2", 300.00)
+
+    updated_customer = cm.Customer.get(cm.Customer.customer_id == 2)
+    assert updated_customer.credit_limit == 300.00
+
+    # Search for a customer
+    target_customer = bo.search_customer(customer4["customer_id"])
+    assert target_customer["email_address"] == customer4["email_address"]
+
+    # List active customers
+    assert bo.list_active_customers() == 1
```

Comment:
Good test, this is the idea for a lifecycle test. I may not have been clear about the differences. I'll fix that next Tuesday. At any rate, I'm not sure I'd bother with integration tests yet, since there isn't much being integrated. When something is actually using the functions in basic_operations, that is probably time for integration tests.